### DISCUSSION TITLE
Questions Regarding Traffic Data, Matrix and Performance in Valhalla

### DISCUSSION THREAD

**Discussion by mlldaniel** [valhalla/valhalla#4790] [General]
Hi,

I am Daniel Hwang from South Korea. I apologize for any language barriers as English isn't my first language. I have some questions regarding the use of traffic data and performance in Valhalla:

1. I came across two conflicting pieces of information. The first is from [this page](https://gis-ops.com/traffic-in-valhalla/#user-content-live-traffic-data), which states that real-time and historical traffic data cannot be used for matrix (many-to-many points) endpoints. However, the [Valhalla documentation](https://valhalla.github.io/valhalla/mjolnir/historical_traffic/) mentions that "Valhalla supports both types of traffic for all APIs, excluding multimodal costing". Could you clarify which one is correct?

2. I am a bit confused about the speed values in Valhalla APIs as mentioned [here](https://valhalla.github.io/valhalla/speeds/). It states that real-time and historical data are not used to calculate speed values. Does this imply that the /route and /isochrone endpoints do not consider the speed of the links?

3. Lastly, is the performance result discussed [here](https://github.com/valhalla/valhalla/issues/2604) still relevant?

Thank you.

Best regards,
Daniel Hwang

**Comment by nilsnolde** [comment]
1. the page you're linking to is > 2 years old, matrix traffic implementation less than that
2. > Does this imply that the /route and /isochrone endpoints do not consider the speed of the links?
    Can you please specify which exact quote makes you believe that? Those docs are vastly outdated (as so much of this kind of docs, we should just remove them..)
3. yes, and as also mentioned there, it will never considerably change. I improved matrix performance by 50-80% since then, but it's nowhere near OSRM and will never be.

**Reply by mlldaniel** [reply]
Thank you for your answer
